compiler/qsc/src/interpret/stateful.rs

```diff
@@ -414,6 +422,76 @@ impl Interpreter {
         Ok(result)
     }
 
+    /// # Errors
+    /// If the currently configured target profile is not `TargetProfile::Base`, an error is returned.
+    /// If the parsing of the expr fails, an error is returned.
+    /// If the compilation of the expr fails, an error is returned.
+    /// If there is a runtime error when generating code for the expr, an error is returned.
+    /// # Panics
+    /// If internal compiler state is inconsistent, a panic may occur.
+    pub fn qirgen(&mut self, expr: &str) -> Result<String, Vec<LineError>> {
+        if self.target != TargetProfile::Base {
+            return Err(vec![LineError(WithSource::new(
+                expr.into(),
+                LineErrorKind::TargetMismatch,
+                None,
+            ))]);
+        }
+
+        let mut codegen = BaseProfGen::new();
+
+        let mut fragment = self.compiler.compile_expr(expr).map_err(|errors| {
+            let source = expr.into();
+            errors
+                .into_iter()
+                .map(|error| LineError(WithSource::new(Arc::clone(&source), error.into(), None)))
+                .collect::<Vec<_>>()
+        })?;
+
+        let pass_errors = self.passes.run(
+            self.store.core(),
+            self.compiler.assigner_mut(),
+            &mut fragment,
+        );
+        if !pass_errors.is_empty() {
+            let source = expr.into();
+            return Err(pass_errors
+                .into_iter()
+                .map(|error| LineError(WithSource::new(Arc::clone(&source), error.into(), None)))
+                .collect());
+        }
+
+        let Fragment::Stmt(stmt) = fragment else {
+            panic!("only a stmt fragment should reach here.")
```

Comment:
Can we make `compile_expr` return the `Stmt` instead? (We can panic in `compile_expr` function if the `Fragment` isn't a statement, but the intent will be much clearer since we created the fragment in that same function)

Reply:
Yeah, I had the same thought at one point, but the desire to reuse the pass code without having to introduce a different, codegen-specific variant led me to keep it as-is.